// @patternfly/react-core/dist/esm/components/Alert/Alert.js
import { __rest } from "tslib";
import * as React from 'react';
import { css } from '@patternfly/react-styles';
import styles from '@patternfly/react-styles/css/components/Alert/alert';
import accessibleStyles from '@patternfly/react-styles/css/utilities/Accessibility/accessibility';
import { AlertIcon } from './AlertIcon';
import { capitalize, getOUIAProps } from '../../helpers';
import { AlertContext } from './AlertContext';
export var AlertVariant;
(function (AlertVariant) {
    AlertVariant["success"] = "success";
    AlertVariant["danger"] = "danger";
    AlertVariant["warning"] = "warning";
    AlertVariant["info"] = "info";
    AlertVariant["default"] = "default";
})(AlertVariant || (AlertVariant = {}));
export const Alert = (_a) => {
    var { variant = AlertVariant.default, isInline = false, isLiveRegion = false, variantLabel = `${capitalize(variant)} alert:`, 'aria-label': ariaLabel = `${capitalize(variant)} Alert`, action = null, title, children = '', className = '', ouiaId } = _a, props = __rest(_a, ["variant", "isInline", "isLiveRegion", "variantLabel", 'aria-label', "action", "title", "children", "className", "ouiaId"]);
    const getHeadingContent = (React.createElement(React.Fragment, null,
        React.createElement("span", { className: css(accessibleStyles.screenReader) }, variantLabel),
        title));
    const customClassName = css(styles.alert, isInline && styles.modifiers.inline, variant !== AlertVariant.default && styles.modifiers[variant], className);
    return (React.createElement("div", Object.assign({}, props, { className: customClassName, "aria-label": ariaLabel }, getOUIAProps('Alert', ouiaId), (isLiveRegion && {
        'aria-live': 'polite',
        'aria-atomic': 'false'
    })),
        React.createElement(AlertIcon, { variant: variant }),
        React.createElement("h4", { className: css(styles.alertTitle) }, getHeadingContent),
        children && React.createElement("div", { className: css(styles.alertDescription) }, children),
        React.createElement(AlertContext.Provider, { value: { title, variantLabel } }, action && (typeof action === 'object' || typeof action === 'string') && (React.createElement("div", { className: css(styles.alertAction) }, action)))));
};
