// @patternfly/react-core/dist/esm/components/Tabs/Tabs.js
import { __rest } from "tslib";
import * as React from 'react';
import styles from '@patternfly/react-styles/css/components/Tabs/tabs';
import buttonStyles from '@patternfly/react-styles/css/components/Button/button';
import { css } from '@patternfly/react-styles';
import AngleLeftIcon from '@patternfly/react-icons/dist/js/icons/angle-left-icon';
import AngleRightIcon from '@patternfly/react-icons/dist/js/icons/angle-right-icon';
import { getUniqueId, isElementInView } from '../../helpers/util';
import { TabButton } from './TabButton';
import { TabContent } from './TabContent';
import { getOUIAProps } from '../../helpers';
export var TabsVariant;
(function (TabsVariant) {
    TabsVariant["div"] = "div";
    TabsVariant["nav"] = "nav";
})(TabsVariant || (TabsVariant = {}));
export class Tabs extends React.Component {
    constructor(props) {
        super(props);
        this.tabList = React.createRef();
        this.scrollLeft = () => {
            // find first Element that is fully in view on the left, then scroll to the element before it
            if (this.tabList.current) {
                const container = this.tabList.current;
                const childrenArr = Array.from(container.children);
                let firstElementInView;
                let lastElementOutOfView;
                let i;
                for (i = 0; i < childrenArr.length && !firstElementInView; i++) {
                    if (isElementInView(container, childrenArr[i], false)) {
                        firstElementInView = childrenArr[i];
                        lastElementOutOfView = childrenArr[i - 1];
                    }
                }
                if (lastElementOutOfView) {
                    container.scrollLeft -= lastElementOutOfView.scrollWidth;
                }
            }
        };
        this.scrollRight = () => {
            // find last Element that is fully in view on the right, then scroll to the element after it
            if (this.tabList.current) {
                const container = this.tabList.current;
                const childrenArr = Array.from(container.children);
                let lastElementInView;
                let firstElementOutOfView;
                for (let i = childrenArr.length - 1; i >= 0 && !lastElementInView; i--) {
                    if (isElementInView(container, childrenArr[i], false)) {
                        lastElementInView = childrenArr[i];
                        firstElementOutOfView = childrenArr[i + 1];
                    }
                }
                if (firstElementOutOfView) {
                    container.scrollLeft += firstElementOutOfView.scrollWidth;
                }
            }
        };
        this.state = {
            shownKeys: [this.props.activeKey] // only for mountOnEnter case
        };
    }
    handleTabClick(event, eventKey, tabContentRef, mountOnEnter) {
        const { shownKeys } = this.state;
        this.props.onSelect(event, eventKey);
        // process any tab content sections outside of the component
        if (tabContentRef) {
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            React.Children.map(this.props.children, (child, i) => {
                child.props.tabContentRef.current.hidden = true;
            });
            // most recently selected tabContent
            tabContentRef.current.hidden = false;
        }
        if (mountOnEnter) {
            this.setState({
                shownKeys: shownKeys.concat(eventKey)
            });
        }
    }
    render() {
        const _a = this.props, { className, children, activeKey, id, isFilled, isSecondary, leftScrollAriaLabel, rightScrollAriaLabel, 'aria-label': ariaLabel, variant, ouiaId, mountOnEnter, unmountOnExit } = _a, props = __rest(_a, ["className", "children", "activeKey", "id", "isFilled", "isSecondary", "leftScrollAriaLabel", "rightScrollAriaLabel", 'aria-label', "variant", "ouiaId", "mountOnEnter", "unmountOnExit"]);
        const { shownKeys } = this.state;
        const uniqueId = id || getUniqueId();
        const Component = variant === TabsVariant.nav ? 'nav' : 'div';
        return (React.createElement(React.Fragment, null,
            React.createElement(Component, Object.assign({ "aria-label": ariaLabel, className: css(styles.tabs, isFilled && styles.modifiers.fill, isSecondary && styles.modifiers.secondary, className) }, getOUIAProps('Tabs', ouiaId), { id: id && id }, props),
                React.createElement("button", { className: css(styles.tabsScrollButton, isSecondary && buttonStyles.modifiers.secondary), "aria-label": leftScrollAriaLabel, onClick: this.scrollLeft },
                    React.createElement(AngleLeftIcon, null)),
                React.createElement("ul", { className: css(styles.tabsList), ref: this.tabList }, React.Children.map(children, (child, index) => {
                    const _a = child.props, { title, eventKey, tabContentRef, id: childId, tabContentId } = _a, rest = __rest(_a, ["title", "eventKey", "tabContentRef", "id", "tabContentId"]);
                    return (React.createElement("li", { key: index, className: css(styles.tabsItem, eventKey === activeKey && styles.modifiers.current, className) },
                        React.createElement(TabButton, Object.assign({ className: css(styles.tabsButton), onClick: (event) => this.handleTabClick(event, eventKey, tabContentRef, mountOnEnter), id: `pf-tab-${eventKey}-${childId || uniqueId}`, "aria-controls": tabContentId ? `${tabContentId}` : `pf-tab-section-${eventKey}-${childId || uniqueId}`, tabContentRef: tabContentRef }, rest), title)));
                })),
                React.createElement("button", { className: css(styles.tabsScrollButton, isSecondary && buttonStyles.modifiers.secondary), "aria-label": rightScrollAriaLabel, onClick: this.scrollRight },
                    React.createElement(AngleRightIcon, null))),
            React.Children.map(children, (child, index) => {
                if (!child.props.children ||
                    (unmountOnExit && child.props.eventKey !== activeKey) ||
                    (mountOnEnter && shownKeys.indexOf(child.props.eventKey) === -1)) {
                    return null;
                }
                else {
                    return React.createElement(TabContent, { key: index, activeKey: activeKey, child: child, id: child.props.id || uniqueId });
                }
            })));
    }
}
Tabs.defaultProps = {
    className: '',
    activeKey: 0,
    onSelect: () => undefined,
    isFilled: false,
    isSecondary: false,
    leftScrollAriaLabel: 'Scroll left',
    rightScrollAriaLabel: 'Scroll right',
    variant: TabsVariant.div,
    mountOnEnter: false,
    unmountOnExit: false
};
