// @patternfly/react-core/dist/js/components/Expandable/Expandable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const React = tslib_1.__importStar(require("react"));
const expandable_1 = tslib_1.__importDefault(require("@patternfly/react-styles/css/components/Expandable/expandable"));
const react_styles_1 = require("@patternfly/react-styles");
const angle_right_icon_1 = tslib_1.__importDefault(require("@patternfly/react-icons/dist/js/icons/angle-right-icon"));
class Expandable extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            isExpanded: props.isExpanded
        };
    }
    calculateToggleText(toggleText, toggleTextExpanded, toggleTextCollapsed, propOrStateIsExpanded) {
        if (propOrStateIsExpanded && toggleTextExpanded !== '') {
            return toggleTextExpanded;
        }
        if (!propOrStateIsExpanded && toggleTextCollapsed !== '') {
            return toggleTextCollapsed;
        }
        return toggleText;
    }
    render() {
        const _a = this.props, { onToggle: onToggleProp, isFocused, isHovered, isActive, className, toggleText, toggleTextExpanded, toggleTextCollapsed, children, isExpanded } = _a, props = tslib_1.__rest(_a, ["onToggle", "isFocused", "isHovered", "isActive", "className", "toggleText", "toggleTextExpanded", "toggleTextCollapsed", "children", "isExpanded"]);
        let onToggle = onToggleProp;
        let propOrStateIsExpanded = isExpanded;
        // uncontrolled
        if (isExpanded === undefined) {
            propOrStateIsExpanded = this.state.isExpanded;
            onToggle = () => {
                onToggleProp();
                this.setState(prevState => ({ isExpanded: !prevState.isExpanded }));
            };
        }
        const computedToggleText = this.calculateToggleText(toggleText, toggleTextExpanded, toggleTextCollapsed, propOrStateIsExpanded);
        return (React.createElement("div", Object.assign({}, props, { className: react_styles_1.css(expandable_1.default.expandable, propOrStateIsExpanded && expandable_1.default.modifiers.expanded, className) }),
            React.createElement("button", { className: react_styles_1.css(expandable_1.default.expandableToggle, isFocused && expandable_1.default.modifiers.focus, isHovered && expandable_1.default.modifiers.hover, isActive && expandable_1.default.modifiers.active), type: "button", "aria-expanded": propOrStateIsExpanded, onClick: onToggle },
                React.createElement(angle_right_icon_1.default, { className: react_styles_1.css(expandable_1.default.expandableToggleIcon), "aria-hidden": true }),
                React.createElement("span", null, computedToggleText)),
            React.createElement("div", { className: react_styles_1.css(expandable_1.default.expandableContent), hidden: !propOrStateIsExpanded }, children)));
    }
}
exports.Expandable = Expandable;
Expandable.defaultProps = {
    className: '',
    toggleText: '',
    toggleTextExpanded: '',
    toggleTextCollapsed: '',
    onToggle: () => undefined,
    isFocused: false,
    isActive: false,
    isHovered: false
};
