// @patternfly/react-core/dist/js/components/ChipGroup/Chip.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const React = tslib_1.__importStar(require("react"));
const react_styles_1 = require("@patternfly/react-styles");
const ChipButton_1 = require("./ChipButton");
const Tooltip_1 = require("../Tooltip");
const times_circle_icon_1 = tslib_1.__importDefault(require("@patternfly/react-icons/dist/js/icons/times-circle-icon"));
const chip_1 = tslib_1.__importDefault(require("@patternfly/react-styles/css/components/Chip/chip"));
const GenerateId_1 = require("../../helpers/GenerateId/GenerateId");
const helpers_1 = require("../../helpers");
class Chip extends React.Component {
    constructor(props) {
        super(props);
        this.span = React.createRef();
        this.renderOverflowChip = () => {
            const { children, className, onClick, ouiaId } = this.props;
            const Component = this.props.component;
            return (React.createElement(Component, Object.assign({ className: react_styles_1.css(chip_1.default.chip, chip_1.default.modifiers.overflow, className) }, helpers_1.getOUIAProps('OverflowChip', ouiaId)),
                React.createElement(ChipButton_1.ChipButton, { onClick: onClick },
                    React.createElement("span", { className: react_styles_1.css(chip_1.default.chipText) }, children))));
        };
        this.renderChip = (randomId) => {
            const { children, closeBtnAriaLabel, tooltipPosition, className, onClick, isReadOnly, ouiaId } = this.props;
            const Component = this.props.component;
            if (this.state.isTooltipVisible) {
                return (React.createElement(Tooltip_1.Tooltip, { position: tooltipPosition, content: children },
                    React.createElement(Component, Object.assign({ className: react_styles_1.css(chip_1.default.chip, isReadOnly && chip_1.default.modifiers.readOnly, className), tabIndex: "0" }, helpers_1.getOUIAProps('Chip', ouiaId)),
                        React.createElement("span", { ref: this.span, className: react_styles_1.css(chip_1.default.chipText), id: randomId }, children),
                        !isReadOnly && (React.createElement(ChipButton_1.ChipButton, { onClick: onClick, "aria-label": closeBtnAriaLabel, id: `remove_${randomId}`, "aria-labelledby": `remove_${randomId} ${randomId}` },
                            React.createElement(times_circle_icon_1.default, { "aria-hidden": "true" }))))));
            }
            return (React.createElement(Component, Object.assign({ className: react_styles_1.css(chip_1.default.chip, isReadOnly && chip_1.default.modifiers.readOnly, className) }, helpers_1.getOUIAProps('Chip', ouiaId)),
                React.createElement("span", { ref: this.span, className: react_styles_1.css(chip_1.default.chipText), id: randomId }, children),
                !isReadOnly && (React.createElement(ChipButton_1.ChipButton, { onClick: onClick, "aria-label": closeBtnAriaLabel, id: `remove_${randomId}`, "aria-labelledby": `remove_${randomId} ${randomId}` },
                    React.createElement(times_circle_icon_1.default, { "aria-hidden": "true" })))));
        };
        this.state = {
            isTooltipVisible: false
        };
    }
    componentDidMount() {
        this.setState({
            isTooltipVisible: Boolean(this.span.current && this.span.current.offsetWidth < this.span.current.scrollWidth)
        });
    }
    render() {
        const { isOverflowChip } = this.props;
        return (React.createElement(GenerateId_1.GenerateId, null, randomId => (isOverflowChip ? this.renderOverflowChip() : this.renderChip(randomId))));
    }
}
exports.Chip = Chip;
Chip.defaultProps = {
    closeBtnAriaLabel: 'close',
    className: '',
    isOverflowChip: false,
    isReadOnly: false,
    tooltipPosition: 'top',
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    onClick: (_e) => undefined,
    component: 'div'
};
