// @patternfly/react-core/dist/js/components/ChipGroup/ChipGroupToolbarItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const React = tslib_1.__importStar(require("react"));
const react_styles_1 = require("@patternfly/react-styles");
const chip_group_1 = tslib_1.__importDefault(require("@patternfly/react-styles/css/components/ChipGroup/chip-group"));
const ChipGroup_1 = require("./ChipGroup");
const ChipButton_1 = require("./ChipButton");
const Tooltip_1 = require("../Tooltip");
const times_icon_1 = tslib_1.__importDefault(require("@patternfly/react-icons/dist/js/icons/times-icon"));
const GenerateId_1 = require("../../helpers/GenerateId/GenerateId");
class ChipGroupToolbarItem extends React.Component {
    constructor(props) {
        super(props);
        this.heading = React.createRef();
        this.state = {
            isTooltipVisible: false
        };
    }
    componentDidMount() {
        this.setState({
            isTooltipVisible: Boolean(this.heading.current && this.heading.current.offsetWidth < this.heading.current.scrollWidth)
        });
    }
    render() {
        const _a = this.props, { categoryName, children, className, isClosable, closeBtnAriaLabel, onClick, tooltipPosition } = _a, rest = tslib_1.__rest(_a, ["categoryName", "children", "className", "isClosable", "closeBtnAriaLabel", "onClick", "tooltipPosition"]);
        if (React.Children.count(children)) {
            const renderChipGroup = (id, HeadingLevel) => (React.createElement("ul", Object.assign({ className: react_styles_1.css(chip_group_1.default.chipGroup, chip_group_1.default.modifiers.toolbar, className) }, rest),
                React.createElement("li", null,
                    this.state.isTooltipVisible ? (React.createElement(Tooltip_1.Tooltip, { position: tooltipPosition, content: categoryName },
                        React.createElement(HeadingLevel, { tabIndex: "0", ref: this.heading, className: react_styles_1.css(chip_group_1.default.chipGroupLabel), id: id }, categoryName))) : (React.createElement(HeadingLevel, { ref: this.heading, className: react_styles_1.css(chip_group_1.default.chipGroupLabel), id: id }, categoryName)),
                    React.createElement("ul", { className: react_styles_1.css(chip_group_1.default.chipGroup) }, children),
                    isClosable && (React.createElement("div", { className: "pf-c-chip-group__close" },
                        React.createElement(ChipButton_1.ChipButton, { "aria-label": closeBtnAriaLabel, onClick: onClick, id: `remove_group_${id}`, "aria-labelledby": `remove_group_${id} ${id}` },
                            React.createElement(times_icon_1.default, { "aria-hidden": "true" })))))));
            return (React.createElement(ChipGroup_1.ChipGroupContext.Consumer, null, (HeadingLevel) => React.createElement(GenerateId_1.GenerateId, null, randomId => renderChipGroup(randomId, HeadingLevel))));
        }
        return null;
    }
}
exports.ChipGroupToolbarItem = ChipGroupToolbarItem;
ChipGroupToolbarItem.defaultProps = {
    categoryName: '',
    children: null,
    className: '',
    isClosable: false,
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    onClick: (_e) => undefined,
    closeBtnAriaLabel: 'Close chip group',
    tooltipPosition: 'top'
};
