// @patternfly/react-core/dist/esm/components/ChipGroup/Chip.js
import * as React from 'react';
import { css } from '@patternfly/react-styles';
import { ChipButton } from './ChipButton';
import { Tooltip } from '../Tooltip';
import TimesCircleIcon from '@patternfly/react-icons/dist/js/icons/times-circle-icon';
import styles from '@patternfly/react-styles/css/components/Chip/chip';
import { GenerateId } from '../../helpers/GenerateId/GenerateId';
import { getOUIAProps } from '../../helpers';
export class Chip extends React.Component {
    constructor(props) {
        super(props);
        this.span = React.createRef();
        this.renderOverflowChip = () => {
            const { children, className, onClick, ouiaId } = this.props;
            const Component = this.props.component;
            return (React.createElement(Component, Object.assign({ className: css(styles.chip, styles.modifiers.overflow, className) }, getOUIAProps('OverflowChip', ouiaId)),
                React.createElement(ChipButton, { onClick: onClick },
                    React.createElement("span", { className: css(styles.chipText) }, children))));
        };
        this.renderChip = (randomId) => {
            const { children, closeBtnAriaLabel, tooltipPosition, className, onClick, isReadOnly, ouiaId } = this.props;
            const Component = this.props.component;
            if (this.state.isTooltipVisible) {
                return (React.createElement(Tooltip, { position: tooltipPosition, content: children },
                    React.createElement(Component, Object.assign({ className: css(styles.chip, isReadOnly && styles.modifiers.readOnly, className), tabIndex: "0" }, getOUIAProps('Chip', ouiaId)),
                        React.createElement("span", { ref: this.span, className: css(styles.chipText), id: randomId }, children),
                        !isReadOnly && (React.createElement(ChipButton, { onClick: onClick, "aria-label": closeBtnAriaLabel, id: `remove_${randomId}`, "aria-labelledby": `remove_${randomId} ${randomId}` },
                            React.createElement(TimesCircleIcon, { "aria-hidden": "true" }))))));
            }
            return (React.createElement(Component, Object.assign({ className: css(styles.chip, isReadOnly && styles.modifiers.readOnly, className) }, getOUIAProps('Chip', ouiaId)),
                React.createElement("span", { ref: this.span, className: css(styles.chipText), id: randomId }, children),
                !isReadOnly && (React.createElement(ChipButton, { onClick: onClick, "aria-label": closeBtnAriaLabel, id: `remove_${randomId}`, "aria-labelledby": `remove_${randomId} ${randomId}` },
                    React.createElement(TimesCircleIcon, { "aria-hidden": "true" })))));
        };
        this.state = {
            isTooltipVisible: false
        };
    }
    componentDidMount() {
        this.setState({
            isTooltipVisible: Boolean(this.span.current && this.span.current.offsetWidth < this.span.current.scrollWidth)
        });
    }
    render() {
        const { isOverflowChip } = this.props;
        return (React.createElement(GenerateId, null, randomId => (isOverflowChip ? this.renderOverflowChip() : this.renderChip(randomId))));
    }
}
Chip.defaultProps = {
    closeBtnAriaLabel: 'close',
    className: '',
    isOverflowChip: false,
    isReadOnly: false,
    tooltipPosition: 'top',
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    onClick: (_e) => undefined,
    component: 'div'
};
